### PR TITLE
[7X] Improve runtime performance for gprecoverseg differential recovery

### PR DESCRIPTION
**Pull request by shrakesh**
Issue: current differential recovery uses rsync checksum(-c) option in
rsync. this option was added to avoid any data loss as the rsync
checksum scans for diff even if the size and timestamp are the same. so -c
option was used to prevent the scenario when the size and time
stamp are the same for a file in src and target but the content is different.
This scenario can happen when some query is running on primary and the
coordinator connection to primary is lost but still, the query is
running. at the same time, the mirror got promoted to primary and there
is some other query running. so in this scenario, there is the possibility
that the file on primary and mirror can be updated at the same time with
the same size content but the content is different. In this case without -c
option the rsync will ignore the file to be transferred from source to
target and can cause a data loss scenario.

Fix: to avoid the above scenario of data loss and improve the performance at
the same time. there is an approach in which instead of syncing all the files
using checksum we can do the checksum for only target files, those
target files can be the files that got updated after the primary got down.
other files can be synced without the checksum option.
for the same, I added a new function sync_updated_files().

we were also syncing pg_wal twice which can be done at the end.
there is no point in syncing the pg_control file as the file will have wrong
information because it will have primary related pg_control info.
also, we are reading checkpoints from backup.label file for wal replay
point.

Updated existing test cases for the above-related changes.

Added new test to support the code changes.
1. unit test to verify success and exception state of the newly added
   function
2.  behave test case to validate the sync of updated file with and
    without tablespace.
    
## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: gpMgmt/sbin/gpsegrecovery.py [hunk 1]
```diff
@@ -232,8 +237,69 @@ def sync_pg_data(self):
...
+
+        if srcPath is None:
+            srcPath = self.recovery_info.source_datadir
+        if dstPath is None:
+            dstPath = self.recovery_info.target_datadir
+        if exclude_list is None:
+            exclude_list = []
+
+        # source files is on primary we need to run the remote query to get the list of updated files.
+        cmdStr = "find {} -type f -newermt '{}'".format(srcPath, self.recovery_info.down_time)
```
Comment (Annu149) [inline gpMgmt/sbin/gpsegrecovery.py:266]: What if lot of files got updated on primary after the mirror failed (which will be the actual scenario). Differential will again take good amount of time to calculate checksum. I think the list of files which need syncing should be an intersection of files updated on primary and mirror both after the downtime. Those are the only files we are concerned about

File: gpMgmt/bin/gppylib/recoveryinfo.py [hunk 2]
```diff
@@ -13,7 +14,7 @@ class RecoveryInfo(object):
     Note: we don't have target hostname, since an object of this class will be accessed by the target host directly
     """
     def __init__(self, target_datadir, target_port, target_segment_dbid, source_hostname, source_port,
-                 source_datadir, is_full_recovery, is_differential_recovery,  progress_file):
+                 source_datadir, is_full_recovery, is_differential_recovery, down_time, progress_file):
```
Comment (Annu149) [inline gpMgmt/bin/gppylib/recoveryinfo.py:17]: Not sure if we really want have downtime in recovery info object, can't we fetch it at later point of time?

**Review by Annu149 — COMMENTED** [review]
If possible, can we add a behave scenario to test the changes?

File: gpMgmt/sbin/gpsegrecovery.py [hunk 3]
```diff
@@ -232,8 +237,69 @@ def sync_pg_data(self):
         cmd = Rsync(name="Sync pg data_dir", srcFile=os.path.join(self.recovery_info.source_datadir, ""),
                     dstFile=self.recovery_info.target_datadir,
                     srcHost=self.recovery_info.source_hostname, exclude_list=rsync_exclude_list,
-                    delete=True, checksum=True, progress=True, progress_file=self.recovery_info.progress_file)
+                    delete=True, checksum=False, progress=True, progress_file=self.recovery_info.progress_file)
+        cmd.run(validateAfter=True)
+
+        self.sync_updated_files(exclude_list=rsync_exclude_list)
```
[the remark this answers is not in the document]
Comment (shrakesh) [reply gpMgmt/sbin/gpsegrecovery.py:243]: for a tablespace, the exclude list has to pass none for the sync_updated_files function, as discussed keeping the same. 
another option is to create a separate function, but it is not needed as the exclude list is populated in only 1 function. keeping the same for now.

File: gpMgmt/sbin/gpsegrecovery.py (hunk 1 above)
Comment (shrakesh) [reply gpMgmt/sbin/gpsegrecovery.py:266]: Had a discussion on this and will filter out the common files, also need to discuss the timestamp keeping fts in mind with Ashwin.

**Comment by ashwinstar** [comment]
As I was thinking on your offline question
"we realized that the primary marked-down timestamp which we are getting from the gp_segment_configuration_history table is the timestamp when the fts is triggered the first time after the primary is down/out of network. so there is a little window in which the files can be updated between the span of primary down and fts triggered.  so to get the actual timestamp of the primary down I feel the offset of fts trigger cycle should be considered.
can you please review the approach and suggest."

come to conclusion that the optimization in this PR
"instead of syncing all the files using checksum we can do the checksum for only target files, those target files can be the files that got updated after the primary got down. other files can be synced without the checksum option."
is not feasible. Let me explain the reason below with scenario.

Simple Scenario:

Differential_Optimization

Critical time point in that scenario - at 9:10:00 same block was modified on primary (wrote 2) and mirror (replayed 1) while connection was fully fine between the two nodes. If mirror goes down and differential recover - this block must be copied over. And only way to detect that is based on checksum comparison between the primary and mirror given there is no timestamp or size difference.

So Fact:
Block on primary may be modified ***anytime*** before primary/mirror down event, along with same time block on mirror replayed with older content. Hence, if mirror goes down and has to be differential recovered new primary content has to be copied over. Or if primary goes down and has to be differential recovered old mirror content has to be copied over. To accomplish this have to perform checksum based comparison due to replay timing being off compared to primary update timing and don't see any way around to avoid the same. Thoughts?
Continuing to think....

**Comment by shrakesh** [comment]
Will work on it later, not an high priority now.

**Pull request closed without merging**